// Control.Monad.Writer.Class/index.js

/**
 *  | This module defines the `MonadWriter` type class and its instances.
 */
"use strict";
var Prelude = require("../Prelude");
var Data_Tuple = require("../Data.Tuple");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Data_Unit = require("../Data.Unit");
var Control_Bind = require("../Control.Bind");
var Data_Function = require("../Data.Function");
var Control_Applicative = require("../Control.Applicative");

/**
 *  | The `MonadWriter w` type class represents those monads which support a monoidal accumulator
 *  | of type `w`.
 *  |
 *  | - `writer` appends a value to the accumulator.
 *  | - `listen` modifies the result to include the changes to the accumulator.
 *  | - `pass` applies the returned function to the accumulator.
 *  |
 *  | An implementation is provided for `WriterT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `writer a mempty = pure a`
 *  | - `do { tell x ; tell y } = tell (x <> y)`
 *  | - `listen (pure a) = pure (Tuple a mempty)`
 *  | - `listen (writer a x) = tell x $> Tuple a x`
 *  |
 */
var MonadWriter = function (__superclass_Control$dotMonad$dotMonad_0, listen, pass, writer) {
    this["__superclass_Control.Monad.Monad_0"] = __superclass_Control$dotMonad$dotMonad_0;
    this.listen = listen;
    this.pass = pass;
    this.writer = writer;
};

/**
 *  | The `MonadWriter w` type class represents those monads which support a monoidal accumulator
 *  | of type `w`.
 *  |
 *  | - `writer` appends a value to the accumulator.
 *  | - `listen` modifies the result to include the changes to the accumulator.
 *  | - `pass` applies the returned function to the accumulator.
 *  |
 *  | An implementation is provided for `WriterT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `writer a mempty = pure a`
 *  | - `do { tell x ; tell y } = tell (x <> y)`
 *  | - `listen (pure a) = pure (Tuple a mempty)`
 *  | - `listen (writer a x) = tell x $> Tuple a x`
 *  |
 */
var writer = function (dict) {
    return dict.writer;
};

/**
 *  | Append a value to the accumulator.
 */
var tell = function (dictMonadWriter) {
    return function ($9) {
        return writer(dictMonadWriter)(Data_Tuple.Tuple.create(Data_Unit.unit)($9));
    };
};

/**
 *  | The `MonadWriter w` type class represents those monads which support a monoidal accumulator
 *  | of type `w`.
 *  |
 *  | - `writer` appends a value to the accumulator.
 *  | - `listen` modifies the result to include the changes to the accumulator.
 *  | - `pass` applies the returned function to the accumulator.
 *  |
 *  | An implementation is provided for `WriterT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `writer a mempty = pure a`
 *  | - `do { tell x ; tell y } = tell (x <> y)`
 *  | - `listen (pure a) = pure (Tuple a mempty)`
 *  | - `listen (writer a x) = tell x $> Tuple a x`
 *  |
 */
var pass = function (dict) {
    return dict.pass;
};

/**
 *  | The `MonadWriter w` type class represents those monads which support a monoidal accumulator
 *  | of type `w`.
 *  |
 *  | - `writer` appends a value to the accumulator.
 *  | - `listen` modifies the result to include the changes to the accumulator.
 *  | - `pass` applies the returned function to the accumulator.
 *  |
 *  | An implementation is provided for `WriterT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `writer a mempty = pure a`
 *  | - `do { tell x ; tell y } = tell (x <> y)`
 *  | - `listen (pure a) = pure (Tuple a mempty)`
 *  | - `listen (writer a x) = tell x $> Tuple a x`
 *  |
 */
var listen = function (dict) {
    return dict.listen;
};

/**
 *  | Read a value which depends on the modifications made to the accumulator during an action.
 */
var listens = function (dictMonadWriter) {
    return function (f) {
        return function (m) {
            return Control_Bind.bind((dictMonadWriter["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Bind.Bind_1"]())(listen(dictMonadWriter)(m))(function (v) {
                return Data_Function.apply(Control_Applicative.pure((dictMonadWriter["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Applicative.Applicative_0"]()))(new Data_Tuple.Tuple(v.value0, f(v.value1)));
            });
        };
    };
};

/**
 *  | Modify the final accumulator value by applying a function.
 */
var censor = function (dictMonadWriter) {
    return function (f) {
        return function (m) {
            return pass(dictMonadWriter)(Control_Bind.bind((dictMonadWriter["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Bind.Bind_1"]())(m)(function (v) {
                return Data_Function.apply(Control_Applicative.pure((dictMonadWriter["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Applicative.Applicative_0"]()))(new Data_Tuple.Tuple(v, f));
            }));
        };
    };
};
module.exports = {
    MonadWriter: MonadWriter, 
    censor: censor, 
    listen: listen, 
    listens: listens, 
    pass: pass, 
    tell: tell, 
    writer: writer
};
