// Control.Monad.Reader.Class/index.js

/**
 *  | This module defines the `MonadReader` type class and its instances.
 */
"use strict";
var Prelude = require("../Prelude");
var Control_Category = require("../Control.Category");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Control_Monad = require("../Control.Monad");
var Control_Bind = require("../Control.Bind");
var Control_Applicative = require("../Control.Applicative");

/**
 *  | The `MonadReader` type class represents those monads which support a global context via
 *  | `ask` and `local`.
 *  |
 *  | - `ask` reads the current context.
 *  | - `local f x` changes the value of the local context during the execution of the action `x`.
 *  |
 *  | An implementation is provided for `ReaderT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `do { ask ; ask } = ask`
 *  | - `local f ask = f <$> ask`
 *  | - `local _ (pure a) = pure a`
 *  | - `local f (do { a <- x ; y }) = do { a <- local f x ; local f y }`
 */
var MonadReader = function (__superclass_Control$dotMonad$dotMonad_0, ask, local) {
    this["__superclass_Control.Monad.Monad_0"] = __superclass_Control$dotMonad$dotMonad_0;
    this.ask = ask;
    this.local = local;
};
var monadReaderFun = new MonadReader(function () {
    return Control_Monad.monadFn;
}, Control_Category.id(Control_Category.categoryFn), Control_Semigroupoid.composeFlipped(Control_Semigroupoid.semigroupoidFn));

/**
 *  | The `MonadReader` type class represents those monads which support a global context via
 *  | `ask` and `local`.
 *  |
 *  | - `ask` reads the current context.
 *  | - `local f x` changes the value of the local context during the execution of the action `x`.
 *  |
 *  | An implementation is provided for `ReaderT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `do { ask ; ask } = ask`
 *  | - `local f ask = f <$> ask`
 *  | - `local _ (pure a) = pure a`
 *  | - `local f (do { a <- x ; y }) = do { a <- local f x ; local f y }`
 */
var local = function (dict) {
    return dict.local;
};

/**
 *  | The `MonadReader` type class represents those monads which support a global context via
 *  | `ask` and `local`.
 *  |
 *  | - `ask` reads the current context.
 *  | - `local f x` changes the value of the local context during the execution of the action `x`.
 *  |
 *  | An implementation is provided for `ReaderT`, and for other monad transformers
 *  | defined in this library.
 *  |
 *  | Laws:
 *  |
 *  | - `do { ask ; ask } = ask`
 *  | - `local f ask = f <$> ask`
 *  | - `local _ (pure a) = pure a`
 *  | - `local f (do { a <- x ; y }) = do { a <- local f x ; local f y }`
 */
var ask = function (dict) {
    return dict.ask;
};

/**
 *  | Read a value which depends on the global context in any `MonadReader`.
 */
var reader = function (dictMonadReader) {
    return function (f) {
        return Control_Bind.bindFlipped((dictMonadReader["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Bind.Bind_1"]())(function ($1) {
            return Control_Applicative.pure((dictMonadReader["__superclass_Control.Monad.Monad_0"]())["__superclass_Control.Applicative.Applicative_0"]())(f($1));
        })(ask(dictMonadReader));
    };
};
module.exports = {
    MonadReader: MonadReader, 
    ask: ask, 
    local: local, 
    reader: reader, 
    monadReaderFun: monadReaderFun
};
